File: OverlayManager.cs
using System.Collections.Generic;

/*
 * Copyright (c) 2018, Adam <Adam@sigterm.info>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
namespace OSRSCache
{
	using OverlayDefinition = OSRSCache.definitions.OverlayDefinition;
	using OverlayLoader = OSRSCache.definitions.loaders.OverlayLoader;
	using OverlayProvider = OSRSCache.definitions.providers.OverlayProvider;
	using Archive = OSRSCache.fs.Archive;
	using ArchiveFiles = OSRSCache.fs.ArchiveFiles;
	using FSFile = OSRSCache.fs.FSFile;
	using Index = OSRSCache.fs.Index;
	using Storage = OSRSCache.fs.Storage;
	using Store = OSRSCache.fs.Store;

	public class OverlayManager : OverlayProvider
	{
		private readonly Store store;
		private readonly IDictionary<int, OverlayDefinition> overlays = new Dictionary<int, OverlayDefinition>();

		public OverlayManager(Store store)
		{
			this.store = store;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public void load() throws java.io.IOException
		public virtual void load()
		{
			Storage storage = store.Storage;
			Index index = store.getIndex(IndexType.CONFIGS);
			Archive archive = index.getArchive(ConfigType.OVERLAY.Id);

			byte[] archiveData = storage.loadArchive(archive);
			ArchiveFiles files = archive.getFiles(archiveData);

			foreach (FSFile file in files.Files)
			{
				OverlayLoader loader = new OverlayLoader();
				OverlayDefinition overlay = loader.load(file.FileId, file.Contents);

				overlays[overlay.id] = overlay;
			}
		}

		public virtual ICollection<OverlayDefinition> Overlays
		{
			get
			{
				// return Collections.unmodifiableCollection(overlays.Values);
				return new List<OverlayDefinition>(overlays.Values);
			}
		}

		public virtual OverlayDefinition provide(int overlayId)
		{
			return overlays[overlayId];
		}
	}

}

File: src/main/java/net/runelite/cache/OverlayManager.cpp
#include "OverlayManager.h"
#include "IndexType.h"
#include "ConfigType.h"

namespace net::runelite::cache
{
	using IOException = java::io::IOException;
	using Collection = java::util::Collection;
	using Collections = java::util::Collections;
	using HashMap = java::util::HashMap;
	using Map = java::util::Map;
	using OverlayDefinition = net::runelite::cache::definitions::OverlayDefinition;
	using OverlayLoader = net::runelite::cache::definitions::loaders::OverlayLoader;
	using OverlayProvider = net::runelite::cache::definitions::providers::OverlayProvider;
	using Archive = net::runelite::cache::fs::Archive;
	using ArchiveFiles = net::runelite::cache::fs::ArchiveFiles;
	using FSFile = net::runelite::cache::fs::FSFile;
	using Index = net::runelite::cache::fs::Index;
	using Storage = net::runelite::cache::fs::Storage;
	using Store = net::runelite::cache::fs::Store;

	OverlayManager::OverlayManager(std::shared_ptr<Store> store) : store(store)
	{
	}

	void OverlayManager::load()
	{
		std::shared_ptr<Storage> storage = store->getStorage();
		std::shared_ptr<Index> index = store->getIndex(IndexType::CONFIGS);
		std::shared_ptr<Archive> archive = index->getArchive(ConfigType::OVERLAY.getId());

		std::vector<signed char> archiveData = storage->loadArchive(archive);
		std::shared_ptr<ArchiveFiles> files = archive->getFiles(archiveData);

		for (auto file : files->getFiles())
		{
			std::shared_ptr<OverlayLoader> loader = std::make_shared<OverlayLoader>();
			std::shared_ptr<OverlayDefinition> overlay = loader->load(file->getFileId(), file->getContents());

			overlays.emplace(overlay->getId(), overlay);
		}
	}

	std::shared_ptr<Collection<std::shared_ptr<OverlayDefinition>>> OverlayManager::getOverlays()
	{
		return Collections::unmodifiableCollection(overlays.values());
	}

	std::shared_ptr<OverlayDefinition> OverlayManager::provide(int overlayId)
	{
		return overlays[overlayId];
	}
}
